docs/design/fabric-v2+/fpc-cc-invocation.puml

```diff
@@ -17,156 +17,179 @@ title Foot Box removed
-        with reportbody = <CH_id, TLCC_inst_id> where
-        TLCC_inst_id as unique (randomly generated) identifier of
-        this particular TLCC __instance__
+CLI -> CLI : r_key <- keyGen()
+note right
+    generate a symmetric key to
+    encrypt the FPC chaincode result
+end note
+
+CLI -> CLI : args_ECC <- encrypt <args, r_key> with Chaincode_EK
+
+end
+
+group invoke
+  CLI -> Peer ++: send proposal <CC_Id, args_enc, ...>
+
+  alt FPC Chaincode is not running
+      note right Peer
+          handled implicitly by the peer which determines
+          when to call run() from external builder.
       end note
-      note over Enclave1
-        Verifies local attestation against TLCC_MR and CH_id
-        from sealed state. Remembers TLCC_inst_id.
+
+      create participant "FPC_stub" as ECC1 order 26
+      Peer -> ECC1 ++: call external builder run()
+      note right
+          NOTE: Any failure in ECC or FPC_Enclave will cause
+          ecc (and docker) dieing so peer will notice it.
+      end note
+
+
+      ECC1 -> ECC1 ++: get sealed credentials
+      note right ECC1
+        FPC_stub has access to peer local
+        storage through external builder.
       end note
       return
-    else
-      |||
 
-      ECC1 -> Enclave1 ++: invoke (proposal(args_ECC))
+      create participant "FPC_stub_enclave" as Enclave1 order 30 #99FF99
+      ECC1 -> Enclave1 ++ : spawn enclave with sealed credentials
+      Enclave1 -> Enclave1 : unseal credentials and restore
+      return
+      return
+  end
+
+      Peer -> ECC1 ++: invoke (proposal(args_enc), CTX)
+      note right
+        Transaction context (CTX) used for all interactions
+        with both, the untrusted ledger und trusted ledger (Ledger_enclave)
+        within the invocation. It ensures that all state accesses
+        are from the same state view (block height).
+      end note
+      ECC1 -> Enclave1 ++: invoke (proposal(args_enc), CTX)
+
+      Enclave1 <-> TLCC1: establish secure channel based on local attestation
 
       Enclave1 -> Enclave1 : verify proposal
 
-      Enclave1 -> TLCC1 ++: verify client identity using MSP
+      Enclave1 -> TLCC1 ++: verify (proposal.creator, CTX)
       note over TLCC1
-      TLCC access is needed to check that
-      client identity comes with a certificate
-      that is a proper root of trust.
+      Ledger_enclave access is needed to check that the creator
+      (client identity) passes MSP validation, that is,
+      it comes with a certificate that is a proper root
+      of trust.
       end note
       return
 
 
-      Enclave1 -> Enclave1 : args <- decrypt args_ECC
+      Enclave1 -> Enclave1 : <args, r_key> <- decrypt args_enc with Chaincode_DK
 
       note right
-        If proposal contains optional Req_TLCC_inst_id parameter,
-        make sure it matches own TLCC_inst_id, otherwise abort.
+        If proposal contains optional Req_ledger_enclave_inst_id parameter,
```

Comment:
This is a good point to address, though it's questionable
1) whether different chaincodes should ever access different TLCC
2) whether the client should define what's appropriate.
At a first glance, my preference would be to implement
* credential-release policies for ERCC, e.g., ERCC gives chaincode credentials to the client only if the chaincode's TLCC mrenclave matches the current one in the registry.
* make the chaincode essentially re-register its updated definition

Reply:
> * whether different chaincodes should ever access different TLCC

The reason for this field afair is more to ensure they use the _same_ tlcc (mrenclave &) instance (the dishonest peer could try to have different tlcc instances with different views for a single cc2cc tx)